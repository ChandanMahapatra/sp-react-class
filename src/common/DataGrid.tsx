import React from 'react';
import { HasId } from './banking-types';
import * as lodash from 'lodash';

export type SortIndicator = '⏫' | '⏬' | '';

export type ColumnConfig = {
  field: string;
  label: string;
  sortIndicator?: SortIndicator;
};

type DataGridProps<T> = {
  columns: ColumnConfig[];
  rows: T[];
  selectHeader?: (field: string) => void;
};

function DataGrid<T extends HasId>({ columns, rows, selectHeader }: DataGridProps<T>) {
  return (
    <table className="table table-striped table-hover">
      <thead>
        <tr>
          {columns.map((column) => (
            <th
              key={column.field}
              onClick={() => selectHeader && selectHeader(column.field)}
              style={{cursor: selectHeader ? 'pointer' : 'inherit',}}
            >
              {column.label}
              <span>&nbsp;{column.sortIndicator}</span>
            </th>
          ))}
        </tr>
      </thead>
      <tbody>
        {rows.map((row) => (
          <tr key={row.id}>
            {columns.map((column) => (
              <td key={column.field}>{lodash.get(row, column.field)}</td>
            ))}
          </tr>
        ))}
      </tbody>
    </table>
  );
}

export default DataGrid;
