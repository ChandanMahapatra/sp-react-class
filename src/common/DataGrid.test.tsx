import React from 'react';
import { render } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import DataGrid, { ColumnConfig } from './DataGrid';

let columns: ColumnConfig[];

beforeEach(() => {
  columns = [
    {
      field: 'foo',
      label: 'Foo',
    },
    {
      field: 'bar',
      label: 'Bar',
    },
  ];
});

test('Did a column render?', () => {
  const { getByText } = render(<DataGrid columns={columns} />);
  const header = getByText(columns[0].label);
  expect(header).toBeInTheDocument();
});

test('Did all columns render?', () => {
  const { getByText } = render(<DataGrid columns={columns} />);

  columns.forEach((column) => {
    // expect(() => getByText(column.label)).not.toThrow();
    getByText(column.label);
  });
});
