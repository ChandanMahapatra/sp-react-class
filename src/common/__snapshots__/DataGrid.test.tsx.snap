// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataGrid snapshot 1`] = `
<div>
  <table
    class="table table-striped table-hover"
  >
    <thead>
      <tr>
        <th
          style="cursor: inherit;"
        >
          First Name
          <span>
             
            
          </span>
        </th>
        <th
          style="cursor: inherit;"
        >
          Last Name
          <span>
             
            
          </span>
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td>
          Jenny
        </td>
        <td>
          Sparks
        </td>
      </tr>
      <tr>
        <td>
          Lucas
        </td>
        <td>
          Trent
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`DataGrid with indicator snapshot 1`] = `
<div>
  <table
    class="table table-striped table-hover"
  >
    <thead>
      <tr>
        <th
          style="cursor: inherit;"
        >
          First Name
          <span>
             
            ⏫
          </span>
        </th>
        <th
          style="cursor: inherit;"
        >
          Last Name
          <span>
             
            
          </span>
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td>
          Jenny
        </td>
        <td>
          Sparks
        </td>
      </tr>
      <tr>
        <td>
          Lucas
        </td>
        <td>
          Trent
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
